### PR TITLE
ATen's emscripten support

### PR DESCRIPTION
**Pull request by sinkingsugar**
Not sure if anybody is interested but I managed to infer a `GRU` fine in `wasm` using ATen's compiled with emscripten. It was quite trivial to fix the configuration.
It also passes most of the tests, specially all scalar tensor tests.

The command line to configure was, but could be simplified:
```
emconfigure cmake -DAT_LINK_STYLE=STATIC -DCAFFE2_CMAKE_BUILDING_WITH_MAIN_REPO=OFF -DCMAKE_C_FLAGS="-Wno-implicit-function-declaration -DEMSCRIPTEN -s DISABLE_EXCEPTION_CATCHING=0" -DCMAKE_CXX_FLAGS="-Wno-implicit-function-declaration -DEMSCRIPTEN -s DISABLE_EXCEPTION_CATCHING=0" -DCMAKE_INSTALL_PREFIX=/home/sugar/aten-wasm ../
```

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by sinkingsugar** [comment]
@ezyang Thanks!

Well, if the code hygiene and cleanness of ATen stays as it is I don't foresee any particular issue that cannot be fixed case by case.

But I'm all in for CI/CD as well and I can investigate it, just not sure if you guys care so much ( would probably slow down some builds/ use build agents resources etc )

I'm gonna automate the build for our project anyway and I'll keep an eye and submit PRs in case it breaks since we are using it.

**Comment by ezyang** [comment]
OK, that's good enough for me.

File: aten/src/ATen/CMakeLists.txt [hunk 1]
```diff
@@ -161,31 +161,31 @@ endif(MKLDNN_FOUND)
...
+ set_property(TARGET sleef PROPERTY FOLDER "dependencies")
+ list(APPEND ATen_THIRD_PARTY_INCLUDE ${CMAKE_BINARY_DIR}/include)
+ link_directories(${CMAKE_BINARY_DIR}/sleef/lib)
+ list(APPEND ATen_CPU_DEPENDENCY_LIBS sleef)
+
+ set(CMAKE_CXX_FLAGS ${OLD_CMAKE_CXX_FLAGS})
+
+ # Set these back. TODO: Use SLEEF_ to pass these instead
+ set(BUILD_SHARED_LIBS ${__aten_sleef_build_shared_libs} CACHE BOOL "Build shared libs" FORCE)
+ set(BUILD_TESTS ${__aten_sleef_build_tests} CACHE BOOL "Build tests" FORCE)
```
Comment (apaszke) [inline aten/src/ATen/CMakeLists.txt]: Can we please avoid changing indentation unnecessarily?

**Comment by apaszke** [comment]
It's awesome you managed to port ATen so easily! On one hand you're right, portability makes it easy to use it on a lot of different platforms, but the last time I checked, our math code was actually *terribly* slow when ran in the browser (and the JS file was quite huge). Are you getting some reasonable results now? What's the size of the JS file?

File: aten/src/ATen/CMakeLists.txt (hunk 1 above)
Comment (sinkingsugar) [reply aten/src/ATen/CMakeLists.txt]: No idea why that happened actually, I must have had it commented out at some point!

**Comment by sinkingsugar** [comment]
@apaszke 

Quality of code is pretty good, nothing convoluted so indeed this is a small patch and of course while I'm not a huge fan of the JS/emscripten world, its evolving nicely so everything works just like another platform. I'm going to test it soon on iOS/nintendo switch/PS4 actually, but I suspect compiling should be easy on those platforms.

You are right, it's big ( 300mb in debug, the .a is a big less then 30mb ) and probably slow ( we just use it for inference so that should be ok ) but still it fits the task for us.
The size in release might be smaller, dead code elimination and so on should help.
Still if you are bound to a `nodejs` only environment it's a better option at least for me.

Anyway it's just a start. :smile:

File: aten/src/ATen/CMakeLists.txt (hunk 1 above)
Comment (sinkingsugar) [reply aten/src/ATen/CMakeLists.txt]: although, I just noticed everything else in that cmake file is indented.

**Comment by dnnagy** [comment]
@sinkingsugar Can You give me a full example of how to build libtorch for wasm from source? I want to run an LSTM model inference in the browser by compiling C++ to wasm and then loading it in javascript. Is that even possible? I don't want to train the model, just evaluate a pre-trained one.

**Comment by sinkingsugar** [comment]
@dnnagy I never built full "libtorch" just c10/ATen, the actual high level layer was re-implemented in nim, see https://github.com/fragcolor-xyz/nimtorch

**Pull request closed without merging**